src/Features/CSharp/Portable/StackTraceExplorer/CSharpStackTraceExplorerService.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.StackTraceExplorer;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.StackTraceExplorer
+{
+    [Shared]
+    [ExportLanguageService(typeof(IStackTraceExplorerService), language: LanguageNames.CSharp)]
+    internal class CSharpStackTraceExplorerService : IStackTraceExplorerService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpStackTraceExplorerService()
+        {
+        }
+
+        public string GetClassMetadataName(string className)
+        {
+            var typeNameSyntax = SyntaxFactory.ParseTypeName(className);
+            if (typeNameSyntax is QualifiedNameSyntax qualifiedNameSyntax)
+            {
+                return GetMetadataNameOfQualifiedNameSyntax(qualifiedNameSyntax);
+            }
+
+            return GetGenericNameWithArity(typeNameSyntax);
+        }
+
+        private static string GetMetadataNameOfQualifiedNameSyntax(QualifiedNameSyntax originalSyntax)
+        {
+            SyntaxNode? syntax = originalSyntax;
+            Stack<string> parts = new();
+
+            while (syntax is not null)
+            {
+                if (syntax is QualifiedNameSyntax qualifiedNameSyntax)
+                {
+                    syntax = qualifiedNameSyntax.Left;
+                    parts.Push(GetGenericNameWithArity(qualifiedNameSyntax.Right));
```

Comment:
this seems super weird.  is this tested?  you're adding teh RHS to the stack first.  Then recursing down the LHS.

Reply:
There are tests for sure, the only space I'm lacking now is if there are multiple generic layers in the class. For example: 

`My.Class<T>.InnerClass<U>` should be `My.Class\`1.InnerClass\`1` which still needs a test